dev/merge_spark_pr.py

```diff
@@ -286,68 +281,145 @@ def resolve_jira_issues(title, merge_branches, comment):
+    '[SPARK-1094] Support MiMa for reporting binary compatibility accross versions.'
+    >>> standardize_jira_ref("[WIP]  [SPARK-1146] Vagrant support for Spark")
+    '[SPARK-1146] [WIP] Vagrant support for Spark'
+    >>> standardize_jira_ref("SPARK-1032. If Yarn app fails before registering, app master stays aroun...")
+    '[SPARK-1032] If Yarn app fails before registering, app master stays aroun...'
+    """
+    # If the string is compliant, no need to process any further
+    if (re.search(r'^\[SPARK-[0-9]{3,6}\] (\[[A-Z0-9_\s,]+\] )+\S+', text)):
+        return text
+    
+    # Extract JIRA ref(s):
+    jira_refs = deque()
+    pattern = re.compile(r'(SPARK[-\s]*[0-9]{3,6})', re.IGNORECASE)
+    while (pattern.search(text) is not None):
+        ref = pattern.search(text).groups()[0]
+        # Replace any whitespace with a dash & convert to uppercase
+        jira_refs.append('[' + re.sub(r'\s+', '-', ref.upper()) + ']')
+        text = text.replace(ref, '')
+
+    # Extract spark component(s):
+    components = deque()
+    # Look for alphanumeric chars, spaces, dashes, periods, and/or commas
+    pattern = re.compile(r'(\[[\w\s,-\.]+\])', re.IGNORECASE)
+    while (pattern.search(text) is not None):
+        component = pattern.search(text).groups()[0]
+        # Convert to uppercase
+        components.append(component.upper())
+        text = text.replace(component, '')
+
+    # Cleanup remaining symbols:
+    pattern = re.compile(r'^\W+(.*)', re.IGNORECASE)
+    if (pattern.search(text) is not None):
+        text = pattern.search(text).groups()[0]
+
+    # Assemble full text (JIRA ref(s), module(s), remaining text)
+    if (len(jira_refs) < 1):
+        jira_ref_text = ""
+    jira_ref_text = ' '.join(jira_refs).strip()
+    if (len(components) < 1):
+        components = ""
+    component_text = ' '.join(components).strip()
+    
+    if (len(jira_ref_text) < 1 and len(component_text) < 1):
+        clean_text = text.strip()
+    elif (len(jira_ref_text) < 1):
+        clean_text = component_text + ' ' + text.strip()
+    elif (len(component_text) < 1):
+        clean_text = jira_ref_text + ' ' + text.strip()
+    else:
+        clean_text = jira_ref_text + ' ' + component_text + ' ' + text.strip()
+    
+    return clean_text
+
+def main():
+    os.chdir(SPARK_HOME)
+    original_head = run_cmd("git rev-parse HEAD")[:8]
+    
+    branches = get_json("%s/branches" % GITHUB_API_BASE)
+    branch_names = filter(lambda x: x.startswith("branch-"), [x['name'] for x in branches])
+    # Assumes branch names can be sorted lexicographically
+    latest_branch = sorted(branch_names, reverse=True)[0]
+
+    pr_num = raw_input("Which pull request would you like to merge? (e.g. 34): ")
+    pr = get_json("%s/pulls/%s" % (GITHUB_API_BASE, pr_num))
+    pr_events = get_json("%s/issues/%s/events" % (GITHUB_API_BASE, pr_num))
+
+    url = pr["url"]
+    title = standardize_jira_ref(pr["title"])
```

Comment:
If the title has been modified, I'd prompt the user here for a yes/no as to whether to use the modified title. (i.e. just say "I've re-written the title as follows to match the format: <old> <new>. Would you like to use the new title?"... with <old> and <new> each on a new line). That way if our heuristic is broken in some way the committer can just ignore the re-writing.